dev/io.openliberty.reporting.internal/src/io/openliberty/reporting/internal/DataCollector.java

```diff
@@ -0,0 +1,165 @@
+	private final String productVersion;
+	private final String productEdition;
+
+	private final Set<String> iFixSet = new HashSet<String>();
+
+	private final String os;
+
+	private final String osArch;
+
+	/**
+	 * <p>
+	 * Collects all the required data.
+	 * </p>
+	 * 
+	 * @param featureProvisioner
+	 * @param FixManager
+	 * @param serverInfo
+	 * @throws IOException
+	 * @throws DataCollectorException
+	 */
+	public DataCollector(FeatureProvisioner featureProvisioner, FixManager FixManager, ServerInfoMBean serverInfo)
+			throws IOException, DataCollectorException {
+
+		Map<String, ? extends ProductInfo> allProductInfo;
+		try {
+			allProductInfo = ProductInfo.getAllProductInfo();
+
+			uniqueID = HashUtils.hashString(serverInfo.getInstallDirectory()); // placeholder.
+
+			javaRuntimeInfo = serverInfo.getJavaRuntimeVersion();
+
+			installedFeatures.addAll(getPublicFeatures(featureProvisioner));
+
+			String javaVM = System.getProperty("java.vendor").toLowerCase();
+
+			if (javaVM == null) {
+				javaVM = System.getProperty("java.vm.name", "unknown").toLowerCase();
+			}
+
+			javaVendor = javaVM;
+
+			iFixSet.addAll(FixManager.getIFixes());
+
+			os = System.getProperty("os.name");
+
+			osArch = System.getProperty("os.arch");
+
+			// the key is the productId
+			if (allProductInfo.containsKey("com.ibm.websphere.appserver")) {
+				productVersion = allProductInfo.get("com.ibm.websphere.appserver").getVersion();
+				productEdition = allProductInfo.get("com.ibm.websphere.appserver").getEdition();
+			} else {
+				productVersion = allProductInfo.get("io.openliberty").getVersion();
+				productEdition = allProductInfo.get("io.openliberty").getEdition();
+			}
+
+		} catch (ProductInfoParseException | DuplicateProductInfoException | ProductInfoReplaceException e) {
+			throw new DataCollectorException("Unable to parse Product Info", e);
+		}
+	}
+
+	/**
+	 * 
+	 * @param featureProvisioner
+	 * @return Set<String>
+	 */
+	private Set<String> getPublicFeatures(FeatureProvisioner featureProvisioner) {
+		Set<String> publicFeatures = new TreeSet<String>();
+		Iterator<String> it = featureProvisioner.getInstalledFeatures().iterator();
+		while (it.hasNext()) {
+			String feature = it.next();
+			FeatureDefinition fd = featureProvisioner.getFeatureDefinition(feature);
+
+			if (fd != null && fd.getVisibility() == Visibility.PUBLIC) {
+				if (fd.getSymbolicName().contains("versionless")) {
+					continue;
+				}
+				// get name from feature definition.
+				// input ones come from the cache which is lower case.
+				// If we don't want to include auto features, then check each feature before
+				// adding it.
+				if (!fd.getFeatureName().contains(":")) {
+					if (fd instanceof ProvisioningFeatureDefinition) {
+						publicFeatures.add(fd.getFeatureName());
+
+					} else {
+						// If we're not an instance of ProvisioningFeatureDefinition then add the
+						// feature to the list.
+						publicFeatures.add(fd.getFeatureName());
+					}
```

Comment:
```suggestion
					publicFeatures.add(fd.getFeatureName());
```

This gets much simpler now.

Reply:
Done.